### PR TITLE
Fixed multiprocessing speed issue

### PR DESCRIPTION
**Pull request by jonn-smith**

### CODE REVIEW

File: src/annmas/annotate/command.py [hunk 1]
```diff
@@ -19,9 +20,22 @@
 
 from ..meta import VERSION
 
+__SLEEP_LEN_S = 0.000001
+__MP_BUFFER_LEN_BYTES = 5 * 1024 * 1024  # 5 Meg buffer
+__MP_MESSAGE_DELIMITER = 0xDEADBEEF
+__SENTINEL_VALUE = r"THE _MOST_ CROMULENT SENTINEL VALUE."
```
Comment (kvg) [inline src/annmas/annotate/command.py:26]: Hah, nice :).

File: src/annmas/annotate/command.py [hunk 2]
```diff
@@ -125,36 +137,38 @@ def main(model, threads, output_bam, input_bam):
...
+        logger.debug("Terminating output process.")
+
         # Now that our input processes are done, we can add our exit sentinel onto the output queue and
         # wait for that process to end:
-        results.put(None)
+        results.put(__SENTINEL_VALUE)
         output_worker.join()
 
     logger.info(f"annmas: {mod_name} finished.")
+    logger.info(f"annmas: elapsed time: %2.5fs.", time.time() - t_start)
```
Comment (kvg) [inline src/annmas/annotate/command.py:168]: Wow!  This is *incredibly* fast now!

$ python -m annmas annotate -t 12 -o test_output.bam test_input.bam | tail -3
[INFO 2021-02-04 22:32:47] annmas annotate: annotated 100 reads with 7203 total sections.
[INFO 2021-02-04 22:32:47] annmas: annotate finished.
[INFO 2021-02-04 22:29:19] annmas: elapsed time: 14.04565s.

**Review by kvg — APPROVED** [review]
Looks awesome!  Seems to be averaging around 10-15 reads/second in testing.  Great work!

**Comment by winni2k** [comment]
Hi everybody!

I was reading this PR with interest and noticed the way that data is being moved around the threads, which is pretty good, but could perhaps be made faster, even. It looks like this code is making several copies of each read: 
There's a copy during conversion to string and copy to get the string into the queue (https://github.com/broadinstitute/annmas/blob/44f2399fdc4d525c51f8c136412fe205378d525e/src/annmas/annotate/command.py#L145-L150) and then possibly a copy to get the string back out of the queue (but probably not, https://github.com/broadinstitute/annmas/blob/44f2399fdc4d525c51f8c136412fe205378d525e/src/annmas/annotate/command.py#L243) and a copy to unpack the string back into a pysam alignment (https://github.com/broadinstitute/annmas/blob/44f2399fdc4d525c51f8c136412fe205378d525e/src/annmas/annotate/command.py#L254). I would not be surprised if this copying is a bottleneck. Unfortunately, this is a limitation of how standard library python handles multiprocessing, and probably cannot be improved without using an external library. 

However, if external libraries are on the table, then can I recommend the [ray library](https://docs.ray.io/en/master/walkthrough.html#objects-in-ray)? I recently had this problem and solved it by using ray, which uses an object store and in-memory wrapping of shared variables for many objects. So for example, for a numpy array, you can get away with just one copy. Unfortunately, I'm not confident it will work with a pysam aligned segment object, but I do recommend taking a look. Ray also abstracts away queues, which simplifies queue syntax substantially.

Best wishes,
Winni

**Comment by kvg** [comment]
Thanks, @winni2k!  Definitely looks like it's worth a look (both trimming what we pass around and investigating the ray library).  @jonn-smith, shall we do that in a separate branch so that we can use the current code to process the current MAS-seq flowcells?

**Comment by jonn-smith** [comment]
@winni2k sounds good!  I'll take a look at it.

@kvg Yeah - let's merge this then add in `ray` after so we have something that works for now.

**Pull request merged**